Use span name table instead of materialized view

llinder opened the pull request:
- This switches span name and service name lookups to use a table instead of a materialized view.  This change will fix #1360
- Change limit logic to limit on a set of trace IDs instead of limiting on the provided collection first. In pracitce this didn't make a noticiable difference in the results but it seems like the intended logic.

llinder committed:
Use span name table instead of materialized view

- This switches span name and service name lookups to use a table instead of a materialized view.  This change will fix #1360
- Change limit logic to limit on a set of trace IDs instead of limiting on the provided collection first. In pracitce this didn't make a noticiable difference in the results but it seems like the intended logic.

codefromthecrypt commented:
merging tomorrow unless someone says not too cc @openzipkin/cassandra

codefromthecrypt reviewed (approved):
might want to check the README to ensure nothing is invalidated there

michaelsembwever commented:
please hold @adriancole 
would like to check this.

codefromthecrypt commented:
@michaelsembwever no problemo

michaelsembwever commented:
This looks good. It's a shame that MV didn't work out, but i can't see a way around it compared to the simplicity of a manually denormalised table.

Am updating the stress yaml files appropriately for the schema change.

Looking into compatibility too. I see no incompatibility, although the MV should be dropped, and make sure the table gets created within an existing keyspace.

michaelsembwever commented:
## Stress results

Very rough runs on a dual-core laptop.
15k partitions entered into each table over 5 minutes, in parallel.
Stress yaml profiles were used from the project according to their documentation.

### Write Only

**Existing**

```
traces
Latency mean              :    2.9 ms [insert: 2.9 ms]
Latency median            :    2.1 ms [insert: 2.1 ms]
Latency 95th percentile   :    5.9 ms [insert: 5.9 ms]
Latency 99th percentile   :   28.5 ms [insert: 28.5 ms]
```

```
trace_by_service_span
Latency mean              :    3.6 ms [insert: 3.6 ms]
Latency median            :    0.8 ms [insert: 0.8 ms]
Latency 95th percentile   :   12.2 ms [insert: 12.2 ms]
Latency 99th percentile   :   43.4 ms [insert: 43.4 ms]

```

**Changes here**

```
traces
Latency mean              :    4.4 ms [insert: 4.4 ms]
Latency median            :    3.0 ms [insert: 3.0 ms]
Latency 95th percentile   :    8.6 ms [insert: 8.6 ms]
Latency 99th percentile   :   35.1 ms [insert: 35.1 ms]
```

```
span_name_by_service
Latency mean              :    2.1 ms [insert: 2.1 ms]
Latency median            :    0.6 ms [insert: 0.6 ms]
Latency 95th percentile   :    3.5 ms [insert: 3.5 ms]
Latency 99th percentile   :   30.7 ms [insert: 30.7 ms]
```

```
trace_by_service_span
Latency mean              :    1.9 ms [insert: 1.9 ms]
Latency median            :    0.8 ms [insert: 0.8 ms]
Latency 95th percentile   :    2.8 ms [insert: 2.8 ms]
Latency 99th percentile   :   29.8 ms [insert: 29.8 ms]
```

### Write&Read

**Existing**

```
traces
Latency mean              :    1.8 ms [by_annotation: 2.0 ms, by_trace: 1.1 ms, by_trace_ts_id: 1.2 ms, insert: 2.9 ms]
Latency median            :    1.0 ms [by_annotation: 1.3 ms, by_trace: 0.5 ms, by_trace_ts_id: 0.5 ms, insert: 2.2 ms]
Latency 95th percentile   :    4.8 ms [by_annotation: 4.5 ms, by_trace: 3.5 ms, by_trace_ts_id: 3.5 ms, insert: 7.4 ms]
Latency 99th percentile   :   14.5 ms [by_annotation: 15.4 ms, by_trace: 10.6 ms, by_trace_ts_id: 12.0 ms, insert: 17.7 ms]
```

```
trace_by_service_span
Latency mean              :    1.6 ms [by_duration: 3.0 ms, insert: 0.9 ms, select: 0.8 ms]
Latency median            :    0.7 ms [by_duration: 2.4 ms, insert: 0.6 ms, select: 0.5 ms]
Latency 95th percentile   :    4.4 ms [by_duration: 6.0 ms, insert: 1.4 ms, select: 1.2 ms]
Latency 99th percentile   :    9.2 ms [by_duration: 15.5 ms, insert: 5.9 ms, select: 4.2 ms]
```

**Changes here**

```
traces
Latency mean              :    3.1 ms [by_annotation: 3.9 ms, by_trace: 1.7 ms, by_trace_ts_id: 1.7 ms, insert: 5.0 ms]
Latency median            :    1.5 ms [by_annotation: 3.1 ms, by_trace: 1.1 ms, by_trace_ts_id: 1.1 ms, insert: 4.6 ms]
Latency 95th percentile   :    7.9 ms [by_annotation: 6.6 ms, by_trace: 3.2 ms, by_trace_ts_id: 3.1 ms, insert: 9.9 ms]
Latency 99th percentile   :   18.9 ms [by_annotation: 18.0 ms, by_trace: 14.2 ms, by_trace_ts_id: 15.2 ms, insert: 23.6 ms]
```

```
span_name_by_service
Latency mean              :    2.0 ms [insert: 1.4 ms, select: 3.1 ms, select_span_names: 1.5 ms]
Latency median            :    1.0 ms [insert: 0.8 ms, select: 2.5 ms, select_span_names: 0.9 ms]
Latency 95th percentile   :    3.7 ms [insert: 2.4 ms, select: 4.6 ms, select_span_names: 2.2 ms]
Latency 99th percentile   :   11.7 ms [insert: 8.7 ms, select: 14.1 ms, select_span_names: 10.1 ms]
```

```
trace_by_service_span
Latency mean              :    2.6 ms [by_duration: 5.0 ms, insert: 1.3 ms, select: 1.5 ms]
Latency median            :    1.0 ms [by_duration: 3.8 ms, insert: 0.8 ms, select: 0.9 ms]
Latency 95th percentile   :    8.0 ms [by_duration: 10.1 ms, insert: 1.5 ms, select: 1.7 ms]
Latency 99th percentile   :   15.3 ms [by_duration: 20.1 ms, insert: 7.4 ms, select: 10.2 ms]

```

michaelsembwever commented:
A concern i've had with the cassandra3 schema is the partition keys of `service_name` in either trace_by_service or span_name_by_service are a bit broad and susceptible to hot spots and wideness.

This problem can be addressed, if need be, on the schema proposed here by re-introducing the `DeduplicatingExecutor` (see the previous cassandra storage code).

michaelsembwever commented:
The stress tests are interesting in that they (potentially) show that the schema proposed here is slower than the existing materialised view. I did not expect this. I wonder if some of the complexity and its latency to the materialised view is hidden from these results. I have no objections to accepting the new schema and any (potentially) small lost in performance.

michaelsembwever commented:
I pushed my commit to this branch containing the updates to the stress yaml profiles.

I didn't actually mean to push it to _your_ fork. I had no idea I could do that… @llinder

codefromthecrypt commented on zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanConsumer.java:
```diff
@@ -190,4 +198,25 @@
+    }
+  }
+
+  static BigInteger traceId(Span span) {
```
needs a rebase on master, as this shifted

codefromthecrypt replied:
basically this is in CassandraUtil now

codefromthecrypt reviewed (changes requested):
Change looks necessary to keep the experience up. We can re-introduce the deduper as a separate change or add it here. If we add it here, let's make sure the test is carried over, too.

just needs a rebase to buff-out the trace id thing.

llinder commented:
@michaelsembwever thanks for updating the stress yaml profiles and running a comparison test.  It is indeed interesting to see the results of the stress test.  Really wish the MV worked as well, especially given it appears to perform better as well.  I will review the `DeduplicatingExecutor` code and get this rebased today or tomorrow at the latest.

Regarding the schema change.  Is the preference to create a new CQL that contains the drop/add statements?

Thanks for the detailed review and comments!

llinder commented:
Rebased the latest upstream changes on master.  Also added the deduper on writes writes on `trace_by_service_span` and `span_name_by_service`.

I think the only outstanding thing to address is the schema changes.  Given that this is still marked _experimental_ do we want to start maintaining schema change sets?  If thats what is preferred I will pull the table/view changes out to a separate file.

michaelsembwever commented:
> I will review the DeduplicatingExecutor code…

Let's leave the DeduplicatingExecutor out for now. 
Feel free to squash my additional commit into your commit, I would appreciate having just one commit in the branch before merging.

> Regarding the schema change. Is the preference to create a new CQL that contains the drop/add statements?

Check out how the Schema class in the old cassandra storage dealt with upgrades.

codefromthecrypt commented:
I don't think we should be dealing with upgrades at this point. One of the reasons we merged earlier was under the assumption that this is experimental. With that in mind, there should be no expectations around schema upgrade logic (that tends to very much clutter the code when in place)

codefromthecrypt commented on zipkin-storage/cassandra3/src/main/java/zipkin/storage/cassandra3/CassandraSpanConsumer.java:
```diff
@@ -162,32 +177,106 @@
     }
   }
+
+  static BigInteger traceId(Span span) {
```
there's still some drift here, but I'll take care of it post merge

codefromthecrypt commented:
@llinder @michaelsembwever @abesto I think there's something wrong with our circleci or something as the cassandra3 tests skip. This led me to merge on false-green. Today, I noticed locally that the tests fail, which is bad, so I reverted this commit and the associated cleanups.

https://github.com/openzipkin/zipkin/commit/10b855231a6736552cf1e01064fdea3afa9704bc

sorry about the confusion. Once it is working locally, please raise another PR. Probably best to start by reverting the revert commit above, so you can start with cleanups applied.

michaelsembwever commented:
> One of the reasons we merged earlier was under the assumption that this is experimental. 

+1

michaelsembwever commented:
I'm not too sure what happened with this merge, but it did not include my commit.

michaelsembwever commented:
Re-created in https://github.com/openzipkin/zipkin/pull/1392
